Adds migration for v1.1.0

leekelleher opened the pull request:
When upgrading to (upcoming) StackedContent (and InnerContent) v1.1.0, the user would need to re-save all the StackedContent data-types, in order to swap over the doc-type aliases with GUIDs.

This migration would automate that step.

> **Note:** This pull request depends on a pull request in the InnerContent repository: https://github.com/umco/umbraco-inner-content/pull/12

---

I'd considered attempting to tackle property-data, but fear took the better of me - the thought of vorto'd-nested-gridified-stacked-contents scared me shitless! 😝

leekelleher commented:
@mattbrailsford I think there's an argument for having both approaches. Maybe the self-healing code isn't required on the data-type, (because the migration handles it - and automates the healing), but for the property-data, that feels like a minefield, (especially with the unknown _crazy-nested-stacked-contents_ that we keep hearing about).

---

My reasoning for taking the migration approach is that I forgot to manually re-save the datatypes when testing the upgrade.  This resulted in a JS error when viewing a node that used a "single-item mode" SC that didn't have a value, _(it may feel edge-case, but it happened during my testing)_.

If we feel this migration is too heavy, then I could look at self-healing the "empty single-item mode" SC via JS? _(I'm sure it's possible)_ - **Should I raise a separate issue for it?** Could help to isolate it?

mattbrailsford commented:
@leekelleher hmm, then yea, that does sound like a bug in the self healing code so would be good to get that resolved.

It's a tricky one as I don't fully know how I feel about including the migrations helper in inner content for use in sc. I'm not sure if that's extending the boundaries beyond what the package's purpose is. Maybe I feel it'd be better as a core PR? Not sure. As I say, i haven't fully formed an opinion atm, these are just my random thoughts :)

leekelleher commented:
re: the migration helper in IC - I did have it all self-contained in SC, but then had that "oh, what if?" moment.  After sleeping on it, I think I'm coming around to investigating the JS self-healing fix.

I'll open a separate issue for it.

leekelleher commented:
I've moved the "Migration Runner" bit back into SC, so we don't need https://github.com/umco/umbraco-inner-content/pull/12 - unless if you think it's a good idea to have it in IC?

Current status is that I've hit 2 snags...

1. Any data-type prevalue changes that are made during a migration, are being reverted by uSync (during app-startup).  For now, I'm manually pausing uSync.
2. With `UpdatePropertyData`, I've switched from doing a direct database update to using the ContentService API, so that there's a log and version history of the change.  BUT something isn't quite right with the RegEx, as any nested values aren't being rendered, SC treats the nested JSON as a string, rather than a JSON object.  From looking at the db, I can't see what the problem is, it may need a fresh pair of eyes? 👀

leekelleher commented:
Closing this PR, the resolution has been superseded by PR #32
(and InnerContent PR https://github.com/umco/umbraco-inner-content/pull/15)
